app/models/assessment.rb

```diff
@@ -2,38 +2,68 @@
 class Assessment < ApplicationRecord
 
   belongs_to :member
-  has_many :questions, inverse_of: :assessment,
-           after_add: :save_fingerprint, after_remove: :save_fingerprint
+  has_many :variants, inverse_of: :assessment,
+           after_add: :save_fingerprint, after_remove: :save_fingerprint,
+           dependent: :destroy
 
   has_many :other_versions,
            -> (a) { where.not(id: a.id) },
            class_name: 'Assessment',
-           foreign_key: :identifier, primary_key: :identifier
-
-  enum visibility: %i[internal external]
+           foreign_key: :a15k_identifier, primary_key: :a15k_identifier
 
   # values are stored here temporarily when they're
   # being sent to and from the metadata repository
   attr_accessor :metadata
 
-  before_validation :set_default_identifier, on: :create
+  before_validation :check_and_initialize_identifiers, on: :create
   before_validation :set_version, on: :create
   before_validation :set_fingerprint
 
-  validates :version, uniqueness: { scope: :identifier }
-  validates :fingerprint, uniqueness: true # { scope: :member_id } # TODO decide if should be scoped
+  validates :a15k_identifier, presence: true
+  validates :a15k_version, presence: true, uniqueness: { scope: :a15k_identifier }
+  validates :source_identifier, presence: true, if: :source_version?
+  validates :source_version, uniqueness: { scope: [:member_id, :source_identifier], allow_blank: true }
+  validates :fingerprint, uniqueness: true
   validate :ensure_member_matches_other_versions
 
   after_commit :send_to_metadata_repo, on: :create
 
   def digest
-    Digest::MD5.hexdigest questions.map(&:content).sort.join
+    Digest::MD5.hexdigest variants.map(&:content).sort.join
   end
 
+  # Put a safety on assessment destruction
+  attr_accessor :enable_destroy
+  before_destroy :abort_unless_destroy_enabled
+
   protected
 
-  def set_default_identifier
-    self.identifier ||= SecureRandom.uuid
+  def check_and_initialize_identifiers
+    if a15k_identifier? && other_versions.empty?
+      errors.add(:a15k_identifier, "can only be given when a prior version has been assigned its value")
+      return
+    end
+
+    if a15k_identifier.present? && source_identifier.present?
+      if other_versions.first.source_identififer != source_identififer
+        errors.add(:source_identifier, "does not match previous values for the given A15k identifier")
+      end
+      return
+    end
+
+    if source_identifier.present?
+      a_version_with_same_source_identifier = self.class.where(
+        member_id: member_id,
+        source_identifier: source_identifier
+      ).first
+
```

Comment:
If we think this would be useful elsewhere, it could be a scope similar to the `other_versions`, maybe  `version_with_same_source_identifier`.

Reply:
I had it as a method at first, but then this was the only use.  So I'd say let's promote it later if the need arises.